### PR TITLE
Implement `parquet` storage

### LINKED ISSUE #1116: Implement `parquet` storage
### [Apache parquet](https://parquet.apache.org/docs/overview/) 

 is a columnar-based binary file format for storing and processing large amounts of data.

 When processing large datasets, it may not be practical to load the entire dataset into memory, so it is more efficient to load and process only parts of the data. 

 To achieve this, columnar-based formats store data by columns rather than by rows, which allows for loading and processing only the required columns for consistency of memory.

  Additionally, columnar-based formats can efficiently compress data, which can save disk space.

 Parquet is used as a file format in distributed storage systems based on Hadoop and is supported in various programming languages.

 Parquet files are commonly used to store large datasets and are a core technology in large-scale data processing systems.

### Hadoop
is a distributed computing framework designed for storing and analyzing large volumes of data.

### PR DESCRIPTION
**Pull request by zmrdltl**
## 🛠️ Background
resolve #1116 

## 🛠️ Reference
[apache arrow web site](https://arrow.apache.org/)
[arrow-rs](https://github.com/apache/arrow-rs)
[parquet crate from arrow-rs](https://docs.rs/parquet/latest/parquet/)

### 🔨 add basic implementation of storage in GlueSQL
- [x] struct ParquetStorage
- [x] impl StoreMut for ParquetStorage
- [x] impl Store for ParquetStorage
- [x] Error

### 🔨  Read Parquet files with sql statement
- [x] fetch_schema
- [x] fetch_all_schemas
- [x] fetch_data
- [x] scan_data

### 🔨  Write Parquet files with sql statement
- [x] insert_schema
- [x] delete_schema
- [x] append_data
- [x] insert_data
- [x] delete_data
- [x] rewrite
- [x] write

### 🔨Tests
parquet create supports read and write configuration settings.
- [x] errors
- [x] all type test

### 🔨docs
- [x] explain storages & usages
- [x] cli examples
- [x] code examples

### 🔨cli
- [x] add promts for parquet-storage

### CODE REVIEW

**Comment by danthegoodman1** [comment]
I’ve recently been looking at using gluesql for a project and noticed this support pending. However I am unsure of (based on how custom storage engines are defined) whether the columnar aspect can be supported: e.g. not all columns need to be returned during a query. Looking at the impl it seems entire rows are returned, rather than just the columns needed

**Comment by gurugio** [comment]
> I’ve recently been looking at using gluesql for a project and noticed this support pending. However I am unsure of (based on how custom storage engines are defined) whether the columnar aspect can be supported: e.g. not all columns need to be returned during a query. Looking at the impl it seems entire rows are returned, rather than just the columns needed

I'm interesting in the columnar as well.
I've heard that it's not easy to apply columnar into the current GluSQL architecture.
I will contact MeenSeek soon and ask him how's it going.

File: storages/parquet-storage/src/store_mut.rs [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+            for row in &rows {
+                match row {
+                    DataRow::Vec(values) => {
+                        let value = values[i].clone();
+
+                        match value {
+                            Value::Null => {
+                                if let ColumnWriter::BoolColumnWriter(ref mut typed) = col_writer {
+                                    typed.write_batch(&[], Some(&[0]), None).map_storage_err()?;
+                                }
```
Comment (panarch) [inline storages/parquet-storage/src/store_mut.rs:207]: we can consider to use match with tuple like this.
e.g.
```rust
let mut col_writer = writer.untyped();
for row in &rows {
    match row {
        DataRow::Vec(values) => {
            let value = values[i].clone();
            let col_writer = &mut col_writer;

            match (value, col_writer) {
                (Value::Null, col_writer) => {
                    //
                }
                (Value::Bool(val), ColumnWriter::BoolColumnWriter(ref mut typed)) => {
                        typed
                            .write_batch(&[val], Some(&[1]), None)
                            .map_storage_err()?;
                }
                _ => {
                    panic!("Invalid value type");
                }
            }
```

File: storages/parquet-storage/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+version.workspace = true
+edition.workspace = true
+description.workspace = true
+license.workspace = true
+repository.workspace = true
+documentation.workspace = true
+
+[dependencies]
+gluesql-core.workspace = true
+utils.workspace = true
```
Comment (panarch) [inline storages/parquet-storage/Cargo.toml:13]: we may remove this. looks not being used

File: storages/parquet-storage/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+async-trait = "0.1"
+serde = { version = "1", features = ["derive"] }
+bincode = "1.3"
+futures = "0.3"
+tokio = { version = "1", features = ["rt", "macros"] }
+hex = "0.4"
```
Comment (panarch) [inline storages/parquet-storage/Cargo.toml:19]: this too

File: storages/parquet-storage/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+thiserror = "1.0"
+parquet = "43.0.0"
+byteorder = "1.4.3"
+num-traits = "0.2.5"
+rust_decimal = "1.15"
+lazy_static = "1.4.0"
+
+[dev-dependencies]
+test-suite.workspace = true
+uuid = "1"
```
Comment (panarch) [inline storages/parquet-storage/Cargo.toml:29]: this too

File: storages/parquet-storage/src/error.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    InvalidParquetContent(String),
+
+    #[error("unmapped parquet type: {0}")]
+    UnmappedParquetType(Type),
+
+    #[error("unmapped parquet converted type: {0}")]
+    UnmappedParquetConvertedType(ConvertedType),
+
+    #[error("unmapped glue data type: {0}")]
+    UnmappedGlueDataType(DataType),
```
Comment (panarch) [inline storages/parquet-storage/src/error.rs:67]: looks not being used.

File: storages/parquet-storage/src/error.rs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    TableNameDoesNotMatchWithFile,
+
+    #[error("invalid parquet file content: {0}")]
+    InvalidParquetContent(String),
+
+    #[error("unmapped parquet type: {0}")]
+    UnmappedParquetType(Type),
+
+    #[error("unmapped parquet converted type: {0}")]
+    UnmappedParquetConvertedType(ConvertedType),
```
Comment (panarch) [inline storages/parquet-storage/src/error.rs:64]: this error type, too.

Comment (panarch) [inline storages/parquet-storage/src/error.rs:64]: these, too.

File: storages/parquet-storage/src/error.rs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+impl From<ParquetStorageError> for Error {
+    fn from(error: ParquetStorageError) -> Self {
+        Self::StorageMsg(error.to_string())
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ParquetStorageError {
+    #[error("unable to set new SerialiszedFileReader")]
+    UnableToSetNewSerializedFileReader,
```
Comment (panarch) [inline storages/parquet-storage/src/error.rs:43]: this too

**Review by panarch — APPROVED** [review]
It's such a pleasure to see this. looks all great!
Let's ship this to the main.

**Comment by zmrdltl** [comment]
> I’ve recently been looking at using gluesql for a project and noticed this support pending. However I am unsure of (based on how custom storage engines are defined) whether the columnar aspect can be supported: e.g. not all columns need to be returned during a query. Looking at the impl it seems entire rows are returned, rather than just the columns needed


Currently, the interface of the plan part is not implemented to read data by column, so the entire row is fully scanned, which requires change. The significance of this work is reading and writing parquet files using sql statements.

**Pull request merged**